p2p/http/libp2phttp.go

```diff
@@ -0,0 +1,688 @@
+	}()
+
+	if len(h.httpTransport.requestedListenAddrs) == 0 && h.streamHost == nil {
+		return ErrNoListeners
+	}
+
+	h.httpTransport.listeners = make([]net.Listener, 0, len(h.httpTransport.requestedListenAddrs)+1) // +1 for stream host
+
+	streamHostAddrsCount := 0
+	if h.streamHost != nil {
+		streamHostAddrsCount = len(h.streamHost.Addrs())
+	}
+	h.httpTransport.listenAddrs = make([]ma.Multiaddr, 0, len(h.httpTransport.requestedListenAddrs)+streamHostAddrsCount)
+
+	errCh := make(chan error)
+
+	if h.streamHost != nil {
+		listener, err := StreamHostListen(h.streamHost)
+		if err != nil {
+			return err
+		}
+		h.httpTransport.listeners = append(h.httpTransport.listeners, listener)
+		h.httpTransport.listenAddrs = append(h.httpTransport.listenAddrs, h.streamHost.Addrs()...)
+
+		go func() {
+			errCh <- http.Serve(listener, &h.rootHandler)
+		}()
+	}
+
+	closeAllListeners := func() {
+		for _, l := range h.httpTransport.listeners {
+			l.Close()
+		}
+	}
+
+	for _, addr := range h.httpTransport.requestedListenAddrs {
+		parsedAddr := parseMultiaddr(addr)
+		// resolve the host
+		ipaddr, err := net.ResolveIPAddr("ip", parsedAddr.host)
+		if err != nil {
+			closeAllListeners()
+			return err
+		}
+
+		host := ipaddr.String()
+		l, err := net.Listen("tcp", host+":"+parsedAddr.port)
+		fmt.Println("HTTPHost.Serve", err)
+		if err != nil {
+			closeAllListeners()
+			return err
+		}
+		h.httpTransport.listeners = append(h.httpTransport.listeners, l)
+
+		// get resolved port
+		_, port, err := net.SplitHostPort(l.Addr().String())
+		if err != nil {
+			closeAllListeners()
+			return err
+		}
+
+		var listenAddr ma.Multiaddr
+		if parsedAddr.useHTTPS && parsedAddr.sni != "" && parsedAddr.sni != host {
+			listenAddr = ma.StringCast(fmt.Sprintf("/ip4/%s/tcp/%s/tls/sni/%s/http", host, port, parsedAddr.sni))
+		} else {
+			scheme := "http"
+			if parsedAddr.useHTTPS {
+				scheme = "https"
+			}
+			listenAddr = ma.StringCast(fmt.Sprintf("/ip4/%s/tcp/%s/%s", host, port, scheme))
+
+		}
+
+		h.httpTransport.listenAddrs = append(h.httpTransport.listenAddrs, listenAddr)
+
+		if parsedAddr.useHTTPS {
+			go func() {
+				srv := http.Server{
+					Handler:   &h.rootHandler,
+					TLSConfig: h.httpTransport.tlsConfig,
+				}
+				errCh <- srv.ServeTLS(l, "", "")
+			}()
+		} else {
+			go func() {
+				errCh <- http.Serve(l, &h.rootHandler)
+			}()
+		}
+	}
+
+	close(h.httpTransport.waitingForListeners)
+	closedWaitingForListeners = true
+
+	expectedErrCount := len(h.httpTransport.listeners)
+	var err error
+	select {
+	case <-h.httpTransport.closeListeners:
+	case err = <-errCh:
+		expectedErrCount--
+	}
+
+	// Close all listeners
+	closeAllListeners()
+	for i := 0; i < expectedErrCount; i++ {
+		<-errCh
+	}
+	close(errCh)
+
+	return err
+}
+
+func (h *HTTPHost) Close() error {
+	close(h.httpTransport.closeListeners)
+	return nil
+}
+
+// SetHttpHandler sets the HTTP handler for a given protocol. Automatically
+// manages the .well-known/libp2p mapping.
```

Comment:
Add an example here. It's not clear if the protocol ID contains a leading / or not, for example.

Reply:
It shouldn't matter. But I agree that an example would be good.